larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ThreeDLinearFitFeatureTool::ThreeDLinearFitFeatureTool() :
+    m_slidingLinearFitWindow(3),
+    m_slidingLinearFitWindowLarge(10000),
+    m_ratioVariables(true),
+    m_addStraightLineLength(true),
+    m_addDiffWithStraightLineMean(true),
+    m_addDiffWithStraightLineSigma(false),
+    m_addDTDLWidth(false),
+    m_addMaxFitGapLength(true),
+    m_addRMSLinearFit(true)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeDLinearFitFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+const pandora::ParticleFlowObject *const pInputPfo)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    ClusterList clusterList;
+    LArPfoHelper::GetTwoDClusterList(pInputPfo, clusterList);
+    float dTdLWidth(0.f),  diffWithStraightLineMean(0.f), diffWithStraightLineSigma(0.f), maxFitGapLength(0.f), rmsSlidingLinearFit(0.f);
+
+    for (const Cluster *const pCluster : clusterList)
+    {
+        float dTdLWidthCluster(-1.f), straightLineLengthLargeCluster(-1.f), diffWithStraightLineMeanCluster(-1.f), diffWithStraightLineSigmaCluster(-1.f),maxFitGapLengthCluster(-1.f), rmsSlidingLinearFitCluster(-1.f);
+
+        this->CalculateVariablesSlidingLinearFit(pCluster, straightLineLengthLargeCluster, diffWithStraightLineMeanCluster, diffWithStraightLineSigmaCluster,dTdLWidthCluster, maxFitGapLengthCluster, rmsSlidingLinearFitCluster);
+
+        if (m_ratioVariables && (straightLineLengthLargeCluster > std::numeric_limits<double>::epsilon()))
+        {
+            diffWithStraightLineMeanCluster  /= straightLineLengthLargeCluster;
+            diffWithStraightLineSigmaCluster /= straightLineLengthLargeCluster;
+            dTdLWidthCluster                 /= straightLineLengthLargeCluster;
+            maxFitGapLengthCluster           /= straightLineLengthLargeCluster;
+            rmsSlidingLinearFitCluster       /= straightLineLengthLargeCluster;
+        }
+
+        diffWithStraightLineMean   += diffWithStraightLineMeanCluster;
+        diffWithStraightLineSigma  += diffWithStraightLineSigmaCluster;
+        dTdLWidth                  += dTdLWidthCluster;
+        maxFitGapLength            += maxFitGapLengthCluster;
+        rmsSlidingLinearFit        += rmsSlidingLinearFitCluster;
+    }
+
+    const int nClusters(clusterList.size());
+    diffWithStraightLineMean   /= nClusters;
```

Comment:
Care here, as dividing a floating point value by an int. This involves an implicit cast, which would be better as const float nClusters(static_cast<float>(clusterList.size()); Do also need to do something in case nClusters is zero.

Reply:
got it